### ISSUE TITLE
Parallelism incorrect on http super pool

### ISSUE THREAD

**Issue by ktoso** [akka/akka-http#125]
<a href="https://github.com/ryandbair"><img src="https://avatars.githubusercontent.com/u/349575?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [ryandbair](https://github.com/ryandbair)**
_Monday Aug 10, 2015 at 18:50 GMT_
_Originally opened as https://github.com/akka/akka/issues/18174_

----

The parallelism value in clientFlow is set as the product of ConnectionPoolSettings' pipeliningLimit and maxConnections[1] which are both defined as being per host. 

superPool only creates one client flow for all connected hosts so the parallelism becomes global to the superPool instead of per host. 

[1] https://github.com/akka/akka/blob/releasing-akka-stream-and-http-experimental-1.0/akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala#L481

**Comment by ktoso** [comment]
<a href="https://github.com/jrudolph"><img src="https://avatars.githubusercontent.com/u/9868?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jrudolph](https://github.com/jrudolph)**
_Tuesday Aug 11, 2015 at 08:20 GMT_

----

I have just glanced over the code but it seems you are right. I wonder if this can be fixed in code or if it's just the documentation / definition of `max-connections` needs to be updated.

/cc @sirthias

**Comment by ktoso** [comment]
<a href="https://github.com/sirthias"><img src="https://avatars.githubusercontent.com/u/135347?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sirthias](https://github.com/sirthias)**
_Wednesday Aug 19, 2015 at 08:43 GMT_

----

@ryandbair Yes, you are right. Things are not quite right currently in this respect.
IMHO the perfect solution would be to use a superpool parallelism factor that is the sum of the `max-connection` settings of all pools that are currently being used by the super-pool flow.
However, this would require dynamic adjustment of the parallelism factor, which is currently not supported by `mapAsync`. We need to set a static parallelism factor a priori, *before* knowing how many pools will actually be used concurrently by the user (via the produced client-flow).

It doesn't look like there is a good heuristic to spare the user from supplying this superpool parallelism herself. I guess we'll have to add an explicit `parallelism` parameter to the `superPool` method.

**Comment by ktoso** [comment]
<a href="https://github.com/ktoso"><img src="https://avatars.githubusercontent.com/u/120979?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ktoso](https://github.com/ktoso)**
_Monday Aug 31, 2015 at 08:27 GMT_

----

Sounds like this can be split into two steps,
1) document current behaviour to not cause confusion 
2) we could make the dynamic version if we get a "dynamic parallelism" mapAsync

WDYT on the 2nd point? // cc @drewhk

**Comment by ktoso** [comment]
<a href="https://github.com/sirthias"><img src="https://avatars.githubusercontent.com/u/135347?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sirthias](https://github.com/sirthias)**
_Monday Aug 31, 2015 at 13:25 GMT_

----

What about adding a new `parallelism` parameter to the `superPool` method?

**Comment by an-tex** [comment]
Hi, this is rather important if you need to make many requests to different hosts but still not overload each host. I like @ktoso idea about the separate parallelism configuration parameter for the superPool in the meantime. Should be a quick fix right?

**Comment by jrudolph** [comment]
Yes, I guess the `superPool` API isn't used as much as the shared connection pool is. It is kind of weird that the shared pool and the individual super pools provide basically the same functionality but use a different kind of logic for this limit.

Introducing another setting would indeed be a quick fix. Unfortunately, we have already misused `max-open-requests` which would have been perfect for that setting...
